proposals/0534-metro-package-exports-support.md

```diff
@@ -0,0 +1,455 @@
+
+### Conditional exports: User conditions
+
+User conditions relates to the ability to set custom condition names within a project, which may be matched by target packages.
+
+Prior art: [Webpack's `resolve.conditionNames`](https://webpack.js.org/configuration/resolve/#resolveconditionnames) and [Node.js' `--conditions` CLI argument](https://nodejs.org/api/cli.html#cli_c_condition_conditions_condition).
+
+**Proposed**:
+
+- P1 (not targeted in first iteration): We will provide a new config option within `metro.config.js` allowing app developers to define additional condition names which statically extend the default conditions matched.
+
+```js
+module.exports = {
+  // ...
+  resolver: {
+    conditionNames: ['browser', 'production'],
+  },
+};
+```
+
+The exact name and shape of this option may change during implementation.
+
+In addition, `resolver.resolveRequest` will continue to provide an escape hatch from Metro's handling of conditional exports, should an app need to override imports from a given package.
+
+### Opting into strict `"exports"` handling
+
+Since we are late to support `"exports"`, we cannot guarantee that enabling spec-compliant behaviour will not break existing users' apps. Even when we assume that packages have enabled `"exports"` via a breaking release, React Native projects may have updated to these under Metro's previous behaviour.
+
+Therefore, Metro will default to lenient and backwards compatible behaviour, as has been noted in previous sections. We will provide an explicit opt-in method for strict handling, placing the responsibility for migration on React Native package authors. This allows these maintainers to have control over how their package behaves and to reconsider the exposed API under `"exports"`.
+
+**Proposed:**
+
+- Provide a strict mode opt-in for packages, enabling spec-compliant behaviour under Metro.
+- Use non-strict handling by default, which will be a non-breaking change for app developers.
+
+The new flag will be exposed as a `reactNative.metroExportsMode` field within `package.json`.
+
+```json
+{
+  "name": "my-package",
+  "exports": {...},
+  // ...
+  "reactNative": {
+    "metroExportsMode": "strict"
+  }
+}
+```
+
+Note: `"reactNative"` (in camel case) will be a distinct field from `"react-native"` (in dash case, used as a `"main"` module alias that we are starting to deprecate with `"exports"`). We are [separately considering a `‌"newArchSupport"` option](https://github.com/microsoft/rnx-kit/issues/1863#issuecomment-1246580474) within this new top-level key, which will encapsulate options available to React Native packages.
```

Comment:
@kelset @cortinico Flagging to continue the discussion from https://github.com/microsoft/rnx-kit/issues/1863.

**History**: We agreed that the suggested `codegenConfig` object in package.json could be generalised, provisionally as `reactNativeConfig`. This new key would be intended to namespace React Native config options and/or future metadata which libraries can configure.

In this PR, I'm positing the wider name of `reactNative` — and seeing what resistance there is 😅.

Additional supporting thought: Under the proposed `"metroExportsMode": "strict"`, including the `"react-native"` field will no longer be allowed by Metro — reducing the likelihood that both of these exist.